pkg/controller/elasticsearch/pdb/reconcile_with_roles.go

```diff
@@ -0,0 +1,440 @@
+	grouped := map[esv1.NodeRole][]int{}
+	visited := make([]bool, n)
+	for _, role := range priority {
+		indices, ok := rolesToIndices[role]
+		if !ok {
+			continue
+		}
+		for _, idx := range indices {
+			if visited[idx] {
+				continue
+			}
+			targetPDBRole := role
+			// if we already assigned a PDB for this role, use that instead
+			if target, ok := roleToTargetPDB[role]; ok {
+				targetPDBRole = target
+			}
+			grouped[targetPDBRole] = append(grouped[targetPDBRole], idx)
+			for _, r := range indicesToRoles[idx].AsSlice() {
+				roleToTargetPDB[esv1.NodeRole(r)] = targetPDBRole
+			}
+			visited[idx] = true
+		}
+	}
+	// transform into the expected format
+	res := make(map[esv1.NodeRole][]appsv1.StatefulSet)
+	for role, indices := range grouped {
+		group := make([]appsv1.StatefulSet, 0, len(indices))
+		for _, idx := range indices {
+			group = append(group, statefulSets[idx])
+		}
+		res[role] = group
+	}
+	return res, nil
+}
+
+// getRolesFromStatefulSet extracts the roles from a StatefulSet's labels.
+func getRolesFromStatefulSet(statefulSet appsv1.StatefulSet) ([]esv1.NodeRole, error) {
+	roles := make([]esv1.NodeRole, 0, len(label.RoleMappings))
+	labels := statefulSet.Spec.Template.Labels
+	if labels == nil {
+		return nil, fmt.Errorf("statefulSet %s/%s has no labels", statefulSet.Namespace, statefulSet.Name)
+	}
+	// if the statefulset is a coordinating role none of the labels are set to true
+	// so we can initially assume it is, and then check for any labels that are set to true
+	// to invalidate the assumption.
+	isCoordinating := true
+
+	// Check each label-role mapping
+	for _, mapping := range label.RoleMappings {
+		if val, exists := labels[mapping.LabelName]; exists && val == "true" {
+			isCoordinating = false
+			roles = append(roles, mapping.Role)
+		}
+	}
+
+	if isCoordinating {
+		return []esv1.NodeRole{esv1.CoordinatingRole}, nil
+	}
+	return roles, nil
+}
+
+// createPDBForStatefulSets creates a PDB for a group of StatefulSets with shared roles.
+func createPDBForStatefulSets(
+	ctx context.Context,
+	es esv1.Elasticsearch,
+	// role is the role used to determine the maxUnavailable value.
+	role esv1.NodeRole,
+	// statefulSets are the statefulSets grouped into this pdb.
+	statefulSets []appsv1.StatefulSet,
+	// allStatefulSets are all statefulSets in the whole ES cluster.
+	allStatefulSets sset.StatefulSetList,
+	meta metadata.Metadata,
+) (*policyv1.PodDisruptionBudget, error) {
+	if len(statefulSets) == 0 {
+		return nil, nil
+	}
+
+	pdb := &policyv1.PodDisruptionBudget{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      esv1.PodDisruptionBudgetNameForRole(es.Name, string(role)),
+			Namespace: es.Namespace,
+		},
+		Spec: buildRoleSpecificPDBSpec(ctx, es, role, statefulSets, allStatefulSets),
+	}
+
+	if es.Spec.PodDisruptionBudget != nil {
+		pdb.Annotations = es.Spec.PodDisruptionBudget.ObjectMeta.Annotations
+		pdb.Labels = es.Spec.PodDisruptionBudget.ObjectMeta.Labels
+	}
```

Comment:
Missing unit test for this? Also the only reason that this does not modify the original maps IIUC is that the following call `meta.Merge` does clone the maps. Maybe at least at a comment?

Reply:
I spoke to you about this, and have a comment that lines up with your thoughts.